File: pattern-02-classical-fallback/skeleton/java/com/csxplore/quantumpatterns/fallback/adapter/FallbackSolver.java
package com.csxplore.quantumpatterns.fallback.adapter;

import com.csxplore.quantumpatterns.fallback.domain.OptimizationRequest;
import com.csxplore.quantumpatterns.fallback.domain.OptimizationResult;
import com.csxplore.quantumpatterns.fallback.port.Solver;

/**
 * Coordinates fallback behaviour.
 *
 * Business services interact only with this component.
 */
public class FallbackSolver implements Solver {

    private final Solver preferredSolver;
    private final Solver fallbackSolver;

    public FallbackSolver(
            Solver preferredSolver,
            Solver fallbackSolver) {

        this.preferredSolver = preferredSolver;
        this.fallbackSolver = fallbackSolver;
    }

    @Override
    public OptimizationResult solve(
            OptimizationRequest request) {

        try {

            return preferredSolver.solve(request);

        } catch (RuntimeException ex) {

            return fallbackSolver.solve(request);

        }

    }

}

File: pattern-02-classical-fallback/skeleton/python/fallback_solver.py
"""
Coordinates fallback execution.
"""

from solver import Solver


class FallbackSolver(Solver):

    def __init__(self, preferred, fallback):
        self.preferred = preferred
        self.fallback = fallback

    def solve(self, request):

        try:
            return self.preferred.solve(request)

        except RuntimeError:
            return self.fallback.solve(request)